library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.block.extensions.api.data;
+
+import org.quiltmc.qsl.base.api.event.ArrayEvent;
+import org.quiltmc.qsl.base.api.event.ParameterInvokingEvent;
+import org.quiltmc.qsl.block.extensions.impl.ExtraBlockDataImpl;
+import org.quiltmc.qsl.block.extensions.impl.QuiltBlockInternals;
+import net.minecraft.block.Block;
+import java.util.HashMap;
+import java.util.Optional;
+
+/**
+ * Represents a {@linkplain BlockDataKey key} to value collection used to store extra data for {@link Block}s.
+ */
+public interface ExtraBlockData {
+	/**
+	 * Constructs a new builder.
+	 *
+	 * @return new builder
+	 */
+	static Builder builder() {
+		return new ExtraBlockDataImpl.BuilderImpl(new HashMap<>());
+	}
```

Comment:
Is this needed? The block's data won't be constructed by the user anyway, you can only add to the builder.